src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs

```diff
@@ -2192,6 +2195,242 @@ static void Main()
+            Assert.Equal(new[] { "Alice1", "Alice2", "Alice3", "Alice4", "Alice5", "Alice6", "Alice7", "Alice8", "Alice9" }, tuple9WithNames.TupleElementNames);
+
+            Assert.Equal(new[] { "System.Int32", "System.String", "System.Int32", "System.String", "System.Int32", "System.String", "System.Int32", "System.String", "System.Int32" },
+                        tuple9WithNames.TupleElementTypes.Select(t => t.ToTestDisplayString()));
+        }
+
+        [Fact]
+        public void CreateTupleTypeSymbol_ElementTypeIsError()
+        {
+            var tupleComp = CreateCompilationWithMscorlib(trivial2uple + trivial3uple + trivalRemainingTuples);
+            var comp = CSharpCompilation.Create("test", references: new[] { MscorlibRef, tupleComp.ToMetadataReference() });
+
+            ITypeSymbol intType = comp.GetSpecialType(SpecialType.System_Int32);
+            ITypeSymbol errorType = comp.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_DynamicAttribute);
+
+            var tupleWithoutNames = comp.CreateTupleTypeSymbol(ImmutableArray.Create(intType, errorType), default(ImmutableArray<string>));
+
+            Assert.Equal(SymbolKind.NamedType, tupleWithoutNames.Kind);
+
+            var types = tupleWithoutNames.TupleElementTypes;
+            Assert.Equal(2, types.Length);
+            Assert.Equal(SymbolKind.NamedType, types[0].Kind);
+            Assert.Equal(SymbolKind.ErrorType, types[1].Kind);
+        }
+
+        [Fact]
+        public void CreateTupleTypeSymbol_BadNames()
+        {
+            var comp = CSharpCompilation.Create("test", references: new[] { MscorlibRef });
+
+            ITypeSymbol intType = comp.GetSpecialType(SpecialType.System_Int32);
+
+            // illegal C# identifiers and blank
+            var tuple2 = comp.CreateTupleTypeSymbol(ImmutableArray.Create(intType, intType), ImmutableArray.Create("123", ""));
+            Assert.Equal(new[] { "123", "" }, tuple2.TupleElementNames);
+
+            // reserved identifiers
+            var tuple3 = comp.CreateTupleTypeSymbol(ImmutableArray.Create(intType, intType), ImmutableArray.Create("return", "class"));
+            Assert.Equal(new[] { "return", "class" }, tuple3.TupleElementNames);
+        }
+
+        [Fact]
+        public void CreateTupleTypeSymbol_VisualBasicElements()
+        {
+            var vbSource = @"Public Class C
+End Class";
+
+            var vbComp = CreateVisualBasicCompilation("VB", vbSource,
+                                compilationOptions: new VisualBasic.VisualBasicCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+
+            vbComp.VerifyDiagnostics();
+
+            var comp = CreateCSharpCompilation("test", "", referencedCompilations: new[] { vbComp }); // no ValueTuple
+
+            ITypeSymbol vbType = (ITypeSymbol)comp.GetMember("C");
+            var tupleWithNames = comp.CreateTupleTypeSymbol(ImmutableArray.Create(vbType, vbType), default(ImmutableArray<string>));
```

Comment:
I would expect this to throw, we shouldn't accept type symbols from other languages.